### PR TITLE
Use the @engine instance variable in TreeManager

### PR DESCRIPTION
**Pull request by cubeguerrero**
The instance variable `@engine` was only removed starting [version 7 of Arel](https://github.com/rails/arel/blob/v7.0.0/lib/arel/tree_manager.rb). For version 6, we still need to use the instance variable.

How to reproduce bug, only happens in development:
1. `zdi zendesk console` (use the latest image)
2. `account = 1.to_account`
3. `account.touch` - will cause an error because the update query is using the shard database.

### CODE REVIEW

**Comment by vanchi-zendesk** [comment]
@cubeguerrero If you don't mind I will take over this as we fix it.

**Comment by vanchi-zendesk** [comment]
This change looks good. But I agree with @grosser. Can we somehow improve the test matrix here to cover for these missing Arel Versions?  How did the tests pass with newer versions without any of these changes?

**Comment by vanchi-zendesk** [comment]
Todo: Open an integration PR with Classic, and add some tests there to ensure it works with sharding properly.

File: lib/arsi/arel_tree_manager.rb [hunk 1]
```diff
@@ -4,23 +4,25 @@ module Arsi
...
-          ::Arel::Visitors::WhereSql.new engine.connection
+          Arel::Table.engine
         end
-        ::Arel::Nodes::SqlLiteral.new viz.accept(@ctx, ::Arel::Collectors::SQLString.new).value
       end
-    else
-      def where_sql
-        return if @ctx.wheres.empty?
-        viz = ::Arel::Visitors::WhereSql.new @engine.connection
-        ::Arel::Nodes::SqlLiteral.new viz.accept @ctx
```
Comment (vanchi-zendesk) [inline lib/arsi/arel_tree_manager.rb]: Are we dropping support for < 6?

**Comment by vanchi-zendesk** [comment]
@grosser Can you take another look please?

File: lib/arsi/arel_tree_manager.rb [hunk 2]
```diff
@@ -4,23 +4,25 @@ module Arsi
...
-        return if @ctx.wheres.empty?
+    def where_sql(provided_engine = :none)
+      return if @ctx.wheres.empty?
 
-        viz = if ::Arel::VERSION[0].to_i > 6
-          ::Arel::Visitors::WhereSql.new(engine.connection.visitor, engine.connection)
+      selected_engine = provided_engine
+      if selected_engine == :none
+        # Arel 9 has no engine reader
+        selected_engine = if self.respond_to?(:engine)
```
Comment (grosser) [inline lib/arsi/arel_tree_manager.rb]: can do a constant assignment like `AREL_ENGINE = version > 9.0` or so to avoid the runtime check and make it easy to delete once we are on arel 9

File: lib/arsi/arel_tree_manager.rb [hunk 3]
```diff
@@ -4,23 +4,25 @@ module Arsi
...
         end
-        ::Arel::Nodes::SqlLiteral.new viz.accept(@ctx, ::Arel::Collectors::SQLString.new).value
       end
-    else
-      def where_sql
-        return if @ctx.wheres.empty?
-        viz = ::Arel::Visitors::WhereSql.new @engine.connection
-        ::Arel::Nodes::SqlLiteral.new viz.accept @ctx
+
+      viz = if ::Arel::Visitors::WhereSql.instance_method(:initialize).arity == 1
```
Comment (grosser) [inline lib/arsi/arel_tree_manager.rb]: same here ... check for the version and cache it 🤷‍♂

File: test/arsi_test.rb [hunk 4]
```diff
@@ -90,6 +91,17 @@
     end
   end
 
+  it "can be enabled" do
+    begin
+      Arsi.disable!
+      Arsi.enable do
+        assert User.where(id: 1, :password => 'hello').delete_all
+      end
```
Comment (grosser) [inline test/arsi_test.rb]: assert that it's disabled after the block

File: lib/arsi/arel_tree_manager.rb (hunk 1 above)
Comment (saturnflyer) [reply lib/arsi/arel_tree_manager.rb]: We already don't test for it. Rails 4 is on 6

File: lib/arsi/arel_tree_manager.rb (hunk 2 above)
Comment (saturnflyer) [reply lib/arsi/arel_tree_manager.rb]: I will. I don't know how to make assumptions about what will change in different versions so I checked for the behavior instead.

File: Gemfile.lock [hunk 5]
```diff
@@ -22,30 +22,23 @@ GEM
...
-    builder (3.2.3)
-    bump (0.5.0)
-    byebug (3.4.0)
-      columnize (~> 0.8)
-      debugger-linecache (~> 1.2)
-      slop (~> 3.6)
-    columnize (0.8.9)
-    concurrent-ruby (1.1.5)
-    debugger-linecache (1.2.0)
+    builder (3.2.4)
```
Comment (mariozaizar) [inline Gemfile.lock]: I wonder if you run bundle update with the `--conservative` mode? 🤔 
Or all of these bumps are required to make it work? or all of them are desired?

File: lib/arsi/arel_tree_manager.rb [hunk 6]
```diff
@@ -2,25 +2,29 @@
...
-        return if @ctx.wheres.empty?
+    def where_sql(provided_engine = :none)
+      return if @ctx.wheres.empty?
 
-        viz = if ::Arel::VERSION[0].to_i > 6
-          ::Arel::Visitors::WhereSql.new(engine.connection.visitor, engine.connection)
+      selected_engine = provided_engine
+      if selected_engine == :none
+        selected_engine = if AREL_WHERE_SQL_ENGINE_ACCESSOR
+          self.engine || Arel::Table.engine
```
Comment (mariozaizar) [inline lib/arsi/arel_tree_manager.rb]: I wouldn't hurt to keep the root namespace on this block 🤔 
```suggestion
          self.engine || ::Arel::Table.engine
```

File: lib/arsi/arel_tree_manager.rb [hunk 7]
```diff
@@ -2,25 +2,29 @@
...
         else
-          ::Arel::Visitors::WhereSql.new engine.connection
+          Arel::Table.engine
```
Comment (mariozaizar) [inline lib/arsi/arel_tree_manager.rb]:
```suggestion
           ::Arel::Table.engine
```

File: lib/arsi/relation.rb [hunk 8]
```diff
@@ -1,6 +1,6 @@
 module Arsi
   module Relation
-    attr_accessor :without_arsi
+    attr_writer  :without_arsi
```
Comment (mariozaizar) [inline lib/arsi/relation.rb]: Am I curious why this change? We won't need to read this attribute anymore?

Comment (mariozaizar) [reply lib/arsi/relation.rb]: Do we need a spec for this kind of change?

File: test/arsi_test.rb [hunk 9]
```diff
@@ -1,4 +1,5 @@
 require_relative "helper"
+SingleCov.covered!
```
Comment (mariozaizar) [inline test/arsi_test.rb]: ❤️

File: lib/arsi/relation.rb [hunk 10]
```diff
@@ -1,6 +1,6 @@
 module Arsi
```
Comment (mariozaizar) [inline lib/arsi/relation.rb]: ⚠️ I'm seeing this warning when running specs:

> [/lib/arsi/relation.rb:17](https://github.com/zendesk/arsi/blob/c7a61cc9776ddff71c6215681985e2ccbcb4d2bd/lib/arsi/relation.rb#L17): warning: instance variable @without_arsi not initialized

File: lib/arsi/relation.rb (hunk 8 above)
Comment (saturnflyer) [reply lib/arsi/relation.rb]: It's defined just below this. I was getting a warning about redefinition of the reader method because we had `attr_accessor :without_arsi` followed immediately by `def with_out_arsi`

File: lib/arsi/relation.rb (hunk 10 above)
Comment (saturnflyer) [reply lib/arsi/relation.rb]: Yeah. I'm not sure the best way. I looked into that but we'd need to tie into the initialization of the object so that the variable exists when it's created. I'll give it another go and see if I can think of something

Comment (mariozaizar) [reply lib/arsi/relation.rb]: Would something like this make sense?
```ruby
    def without_arsi?
      defined?(@without_arsi) || !arsi_scopeable?
    end
```

Comment (vanchi-zendesk) [reply lib/arsi/relation.rb]: This is actually a bug. It should be:

```
    def without_arsi?
      return @without_arsi if defined?(@without_arsi) && @without_arsi
      !arsi_scopeable?
    end
```
